### PR TITLE
Tweak webbackend API to handle operations with connections

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What
From discussions with @Jamakase, we probably need to tweak the web backend API to provide a better UI/UX flow when implementing #3235.

Connections and Operations API calls should be bundled more together using the web backend routes.

## How
*Describe the solution*
- Add check and check_for_update on the operations route to validate if operation configurations are somehow valid before actually creating/updating them
- Bundle updates to connection and operations in a single API route with the WebBackEndConnectionUpdate (needed when we want to attach operations to a connection)
- Mirror and make a WebBackEndConnectionCreate route too

## Pre-merge Checklist
- [x] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. `airbyte-api/src/main/openapi/config.yaml`
1. `airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java`
1. `airbyte-server/src/main/java/io/airbyte/server/handlers/OperationsHandler.java`
1. the rest

### CODE REVIEW

File: airbyte-api/src/main/openapi/config.yaml [hunk 1]
```diff
@@ -972,6 +1014,22 @@ paths:
           description: Connection not found
         "422":
           $ref: "#/components/responses/InvalidInput"
+  /v1/operations/list_unassigned:
```
Comment (Jamakase) [inline airbyte-api/src/main/openapi/config.yaml:1017]: Are you going to implement it for UI needs or for some other use-cases? I do not think that we are going to use it on UI according our current mockups.

Comment (ChristopheDuong) [reply airbyte-api/src/main/openapi/config.yaml:1017]: let's remove it

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java [hunk 2]
```diff
@@ -251,13 +264,64 @@ public ConnectionRead webBackendUpdateConnection(WebBackendConnectionUpdate webB
...
+  }
+
+  private List<UUID> updateOperations(WebBackendConnectionUpdate webBackendConnectionUpdate)
+      throws JsonValidationException, ConfigNotFoundException, IOException {
+    final ConnectionRead connectionRead =
+        connectionsHandler.getConnection(new ConnectionIdRequestBody().connectionId(webBackendConnectionUpdate.getConnectionId()));
+    final List<UUID> originalOperationIds = new ArrayList<>(connectionRead.getOperationIds());
+    final List<UUID> operationIds = new ArrayList<>();
+    for (var operationUpdate : webBackendConnectionUpdate.getWithOperations()) {
+      if (!originalOperationIds.contains(operationUpdate.getOperationId())) {
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:283]: you're trying to handle the case where the are new operations added during the update, right? would it make sense to just check if they have ids or not? the new ones i assume don't have ids yet. 

also (assuming the new ones don't have ids) don't actually be valid `OperationUpdate` objects, i don't think. I think you might need to add an extra struct that's `OperationCreateOrUpdate` where the id is optional. I think that would make things a bit clearer. The way it is now is a little hard to follow and it means that the required fields in the structs aren't respected. there's probably a better idea than what i suggested, but definitely whatever you pick it would nice for it to be a little clearer.

File: airbyte-api/src/main/openapi/config.yaml [hunk 3]
```diff
@@ -1953,6 +2025,10 @@ components:
           properties:
             withRefreshedCatalog:
               type: boolean
+            withOperations:
```
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:2028]: i don't think you need the "with" here. It was added on the other field because it's a boolean so it is trying to use "with" to make that clear.

File: airbyte-api/src/main/openapi/config.yaml [hunk 4]
```diff
@@ -906,6 +906,27 @@ paths:
           description: Connection not found
         "422":
           $ref: "#/components/responses/InvalidInput"
+  /v1/operations/check:
```
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-api/src/main/openapi/config.yaml:909]: I'm going to assume there's some use case for this that I don't know about, It's not clear to me why we need _both_  `/v1/operations/check` and `/v1/operations/check_for_update`. Can't we always just create an `OperationCreate` object from either an `OperationCreate` or an `OperationUpdate`? The only reason I can think of needing the update version is if there is some limited number of transformations that are allowed on an existing operation that don't apply when you're creating one from scratch.

[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-api/src/main/openapi/config.yaml:909]: Maybe you used this as a pattern?

```
  /v1/sources/check_connection:
    post:
      tags:
        - source
      summary: Check connection to the source
      operationId: checkConnectionToSource
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/SourceIdRequestBody"
        required: true
      responses:
        "200":
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/CheckConnectionRead"
        "404":
          description: Source not found
        "422":
          $ref: "#/components/responses/InvalidInput"
  /v1/sources/check_connection_for_update:
    post:
      tags:
        - source
      summary: Check connection for a proposed update to a source
      operationId: checkConnectionToSourceForUpdate
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/SourceUpdate"
        required: true
      responses:
        "200":
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/CheckConnectionRead"
        "404":
          description: Source not found
        "422":
          $ref: "#/components/responses/InvalidInput"
```

if so, this case is a bit different. the case being modeled in the source case is when the caller doesn't have access to the whole source object just the id but in the latter case is trying to check a source that doesn't exist yet. i'm not even sure this pattern makes too much sense in the source case, i think it might have just happened because the one that takes just a source id happened first. we could probably just say you need to send the whole object every time.

**Review by cgardens — APPROVED** [review]
The new webackend api stuff looks good! Couple small comments.

I feel like I don't have a good understanding of what `/v1/operations/check:` and `/v1/operations/check_for_update:` are going to be used for. Did this come up separately from the discussion that you me, and, artem had? Assuming they are used for something that I missed, I think you probably only need one of them.

Definitely a few changes to be made here but I think they are straight forward so going for an approve.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/OperationsHandler.java [hunk 5]
```diff
@@ -144,6 +175,40 @@ public OperationRead getOperation(OperationIdRequestBody operationIdRequestBody)
...
+      operationIds.remove(operationId);
+      boolean sharedOperation = false;
+      for (StandardSync sync : allStandardSyncs) {
+        // Check if other connections are using the same operation
+        if (sync.getConnectionId() != standardSync.getConnectionId() && sync.getOperationIds().contains(operationId)) {
+          sharedOperation = true;
+          break;
+        }
+      }
+      if (!sharedOperation) {
```
Comment (jrhizor) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/OperationsHandler.java:204]: Depending on where the UI goes in the future we may have to change the behavior for deleting connections to not delete even unshared operations, but this should be fine for now.

**Review by jrhizor — APPROVED** [review]
looks good % Charles's comments

File: airbyte-api/src/main/openapi/config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChristopheDuong) [reply airbyte-api/src/main/openapi/config.yaml:909]: The check route is to be used from the UI as artem was mentioning a need to validate before actually creating an operation

For now, it's really simple, but we could extend the check to validate further the parameters of the operation in the future.

For example if you provide a git_url, checking if it's valid and we're able to pull/clone from it etc
